### PR TITLE
Core/Pooling: Fixed less and less objects from pools being spawned the longer the server is running. Port from TrinityCore.

### LINKED ISSUE #5523: Gameobject pooling is broken [$20]
## CURRENT BEHAVIOUR:

- I am currently working on multiple areas adding chests/herbs/nodes to gameobject pools and there seems to be a problem.
- If you make an pool size of 5 for example, sometimes only 4 nodes get spawned, sometimes all 5 get spawned.
- This has been reported here as well: https://github.com/azerothcore/azerothcore-wotlk/pull/4947

## EXPECTED BLIZZLIKE BEHAVIOUR:

- If you set the gameobject pool to 5 then 5 objects should spawn, not less.

## STEPS TO REPRODUCE THE PROBLEM:

1. Start your server.
2. Make sure #5522 is already merged, if not try this bug with this PR because its easier.
3. Check those 6 guids with ".go ob $guide", they are added to a pool of 5: 33304, 33206, 33207, 33200, 33193, 33208

## AC HASH/COMMIT:

https://github.com/azerothcore/azerothcore-wotlk/commit/0a571973d74554eaa5b0ac4e9d4468844a477ebe

## OPERATING SYSTEM:

Ubuntu 20.04

## CUSTOMIZATIONS:

<bountysource-plugin>

---
There is a **[$20 open bounty](https://www.bountysource.com/issues/98143858-gameobject-pooling-is-broken?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
## Changes Proposed:
-  Ported from TrinityCore fixes related to pooling:
https://github.com/TrinityCore/TrinityCore/commit/de80cd2e0de470522a734e8e7a65e81b82bfc5fb
https://github.com/TrinityCore/TrinityCore/commit/c50aefac963aca5daeacf409deca9692c6ca3924
https://github.com/TrinityCore/TrinityCore/commit/8c0a315734840f99455026ef443a21ae73d7bc3b

## Issues Addressed:
- Closes #5523

## Tests Performed:

- Tested ingame with #5522

## How to Test the Changes:
 More info #5523

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Kitzunu** [comment]
```
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Pools/PoolMgr.cpp:296:12: error: unused variable ‘lastDespawned’ [-Werror=unused-variable]
     uint32 lastDespawned = 0;
            ^~~~~~~~~~~~~
```

good pr!

**Comment by UltraNix** [comment]
@Kitzunu Done.

**Comment by Sedatyf** [comment]
So I checked it with the help of @Hacki95 and everything looks great. 
I checked and 5 out of 6 chests are spawned like they should
I looted them to make them disappear and I checked after respawn time, and there is always 5 out of 6 chests spawn at a time. 
And spawned chests are random, it's not always the same.
Congrats on that PR @UltraNix ✔️

**Review by Hacki95 — APPROVED** [review]

**Comment by Footman** [comment]
tested on live. everything fine

File: src/common/Containers.h [hunk 1]
```diff
@@ -64,33 +64,63 @@ namespace acore
 
     namespace Containers
     {
-        template<class T>
-        void RandomResizeList(std::list<T>& list, uint32 size)
+        // replace with std::size in C++17
```
Comment (FrancescoBorzi) [inline src/common/Containers.h:67]: we already use C++17 so this can be replaced now, no need to postpone for later

Comment (Winfidonarleyan) [reply src/common/Containers.h:67]: https://github.com/azerothcore/azerothcore-wotlk/pull/5454

**Comment by Footman** [comment]
breaks compile

File: src/common/Containers.h (hunk 1 above)
Comment (FrancescoBorzi) [reply src/common/Containers.h:67]: @Winfidonarleyan where ?

**Comment by necropola** [comment]
@UltraNix What is the reason for those commented out code segments
```
/*obj.guid == triggerFrom ||*/
```
This is different from how it looks in TC.

**Comment by UltraNix** [comment]
@necropola It was just simple porting from TC. As they made another commit, you can port it as well to AC.

**Comment by necropola** [comment]
@UltraNix yep, I can do this. Looks pretty straightforward, i. e. probably not much I need to adjust in order to apply followup PR to AC, too.

**Pull request merged**